### ISSUE TITLE
array.difference() evaluates argument once per item in array

### ISSUE THREAD

**Issue by dylansmith** [rethinkdb/rethinkdb#6326]
We found an issue with an incredibly long-running query, and traced it to the use of `difference()`.

```
r.db('blah').table('foo')
  .getAll([ '123' ], { index: 'idx__companyId' })
  .getField('id')
  .difference(
    r.db('blah').table('foo_assets')
    .getAll([ '123' ], { index: 'idx__companyId' })
    .getField('foo_id')
  )
```

Now, the docs for `difference()` seems to imply that it's a simply array comparison function, but it seems like the inner query is run for every item in the outer array. We found this out the hard way, for a case where the outer query returns 18,177 items, and the inner query returns 17,005. Yep, 309,099,885 reads.

Is this how `difference()` is supposed to work? I would expect the inner query to be run only once, and its result used as the array argument to `difference()` but it seems this is not the case.

**Comment by marshall007** [comment]
Yes, this is the intended behavior. While this particular example might easily be optimized (if RethinkDB had a query optimization layer), it's not so easy in the general case. This operation is non-atomic, so each time the `difference` is evaluated it's possible that the inner `getAll` could return different results. In that sense, evaluating the inner query on each iteration is the "correct" behavior.

You can optimize this example by only scanning the table once:

```js
r.db('blah').table('foo')
.getAll([ '123' ], { index: 'idx__companyId' })
.fold({ ids: [], foo_ids: [] }, (acc, row) => {
  return {
    ids: acc('ids').append(row('id')),
    foo_ids: acc('foo_ids').append(row('foo_id'))
  }
})
.do(acc => {
  return acc('ids').difference(acc('foo_ids'));
})
```

**Comment by marshall007** [comment]
Oops, I didn't notice your inner query is on a separate table, `foo_assets`. In that case I would do something like this:

```js
r.db('blah').table('foo')
.getAll([ '123' ], { index: 'idx__companyId' })
.getField('id').coerceTo('array')
.do(ids => {
  return r.db('blah').table('foo_assets')
  .getAll([ '123' ], { index: 'idx__companyId' })
  .getField('foo_id')
  .difference(ids);
})
```

This ensures that both `getAll` queries are only evaluated once.

**Comment by dylansmith** [comment]
Ok, thanks for the explanation. I think it would be useful to update the docs so others don't make the same mistake?

**Comment by srh** [comment]
There's no reason why `difference` should evaluate the second argument multiple times.  What does an operation being non-atomic have to do with it?  Should arithmetic operations evaluate their arguments multiple times too?

It's just a pure bug, a badly implemented rewrite-term.  The implementation is also O(n*m) for no reason, it could be O((n+m)*log(m)).

**Comment by srh** [comment]
Note that its current behavior lets the first argument be a sequence that gets streamed -- `X.difference(Y)` transforms into `X.filter(row => Y.contains(row).not())`.  This has the effect of using O(1) memory even if you have two big tables.  But it's still crap because even if you don't want to use an auxiliary data structure, it should operate one batch at a time, not one element at a time.  (There should be one Y query for every 100k elements of X, or whatever.)

What you want if Y is small enough, or if it's already an array, is for it to be sorted and then accessed in log(m) time for each element of X.  If Y is too big for that, honestly it still should (if X is also big) be externally sorted, but maybe first you'd want that for other stuff.

Or if the docs say both should be arrays, then both should be converted to arrays, forget about streaming.

**Comment by thelinuxlich** [comment]
This is interesting, I wonder if there are other ReQL terms with this behavior.

**Comment by srh** [comment]
Likely suspects are every subclass of `rewrite_term_t`.

**Comment by marshall007** [comment]
@srh I didn't realize the docs only mentioned arrays. In that case you're correct, it's just a bug. I had just assumed `.difference()` was intended to behave the way it does now where you can evaluate a function with the current row context.

**Comment by thelinuxlich** [comment]
So it looks like every reql term that transforms its argument to function will reexecute the subquery for each iteration....maybe this can be a source of bad performance in my app, gonna check this out...

**Comment by marshall007** [comment]
@thelinuxlich yea exactly, I initially glossed over the behavior we're seeing with `.difference()` without thinking about it because it's consistent with other terms.

@srh what are your thoughts on how we should go about fixing this? Obviously `.difference()` in particular should only evaluate its term arguments once, but could we apply a more general optimization? I see two (not necessarily mutually exclusive) options:

1. If we can determine that a given subquery is static (does not reference the current row context), rewrite it into a `.do()`.
1. Cache subquery results, which has potential to improve performance for both static and dynamic subqueries.

I think the effort involved in something like (2) is probably rather substantial, but I'd still like to hear your take on it.

**Comment by srh** [comment]
I think r.difference should take arrays as documented.

You can't generally pull a static subquery out into an r.do() because the query might never get executed.  Caching subquery results is a pessimization for a lot of stuff so I wouldn't want to blindly do it.
